### ISSUE TITLE
JsonJacksonCodec cannot deserialize EnumMap

### ISSUE THREAD

**Issue by earmenda** [redisson/redisson#1293]
Hi,

I opened a ticket on the jackson databind project but I think this issue could use some visibility here. 

https://github.com/FasterXML/jackson-databind/issues/1932

The code below can replicate the exception I am seeing. I ran into this when calling operations on  a RScoredSortedSet, but the code below is stripped to just the Codec.

```java
import java.util.EnumMap;

import org.redisson.codec.JsonJacksonCodec;

import io.netty.buffer.Unpooled;

public class JacksonIssueTest {

    public static class EventTest {

        private EnumMap<EnumMapTestEnum, String> enumMapTest;

        public EventTest() {
        }

        public EventTest(EnumMap<EnumMapTestEnum, String> enumMapTest) {
            this.enumMapTest = enumMapTest;
        }

        public EnumMap<EnumMapTestEnum, String> getEnumMapTest() {
            return enumMapTest;
        }

        public void setEnumMapTest(EnumMap<EnumMapTestEnum, String> enumMapTest) {
            this.enumMapTest = enumMapTest;
        }
    }

    public enum EnumMapTestEnum {
        A, B, C;
    }

    public static void main(String args[]) throws Exception {

        EnumMap<EnumMapTestEnum, String> enumMap = new EnumMap(EnumMapTestEnum.class);
        enumMap.put(EnumMapTestEnum.A, "Test");
        EventTest eventTest = new EventTest(enumMap);

        byte[] bytes = JsonJacksonCodec.INSTANCE.getValueEncoder().encode(eventTest);
        JsonJacksonCodec.INSTANCE.getValueDecoder().decode(Unpooled.wrappedBuffer(bytes), null);
    }
}
```

<pre>
Exception in thread "main" com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of java.util.EnumMap out of FIELD_NAME token
 at [Source: io.netty.buffer.ByteBufInputStream@229c6181; line: 1, column: 135] (through reference chain: JacksonIssueTest$EventTest["enumMapTest"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:270)
	at com.fasterxml.jackson.databind.DeserializationContext.reportMappingException(DeserializationContext.java:1247)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1122)
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1075)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._deserializeFromEmpty(StdDeserializer.java:892)
	at com.fasterxml.jackson.databind.deser.std.EnumMapDeserializer.deserialize(EnumMapDeserializer.java:130)
	at com.fasterxml.jackson.databind.deser.std.EnumMapDeserializer.deserialize(EnumMapDeserializer.java:17)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:129)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97)
	at com.fasterxml.jackson.databind.deser.std.EnumMapDeserializer.deserializeWithType(EnumMapDeserializer.java:182)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:502)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:111)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:276)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:178)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:150)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:129)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromAny(AsPropertyTypeDeserializer.java:182)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserializeWithType(UntypedObjectDeserializer.java:553)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:63)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3814)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2924)
	at org.redisson.codec.JsonJacksonCodec$2.decode(JsonJacksonCodec.java:81)
	at JacksonIssueTest.main(JacksonIssueTest.java:40)
</pre>

I believe this is a bug but I apologize if I am misusing the library.

**Comment by earmenda** [comment]
I attempted to circumvent this issue by overriding initTypeInclusion and using type inclusion As.WRAPPER_ARRAY. This fixed my test but I am still seeing some issues with the Redisson client using this approach. 

```java
 @Override
    protected void initTypeInclusion(ObjectMapper mapObjectMapper) {
        TypeResolverBuilder<?> mapTyper = new ObjectMapper.DefaultTypeResolverBuilder(ObjectMapper.DefaultTyping.NON_FINAL) {
            public boolean useForType(JavaType t) {
                switch (_appliesFor) {
                    case NON_CONCRETE_AND_ARRAYS:
                        while (t.isArrayType()) {
                            t = t.getContentType();
                        }
                        // fall through
                    case OBJECT_AND_NON_CONCRETE:
                        return (t.getRawClass() == Object.class) || !t.isConcrete();
                    case NON_FINAL:
                        while (t.isArrayType()) {
                            t = t.getContentType();
                        }
                        // to fix problem with wrong long to int conversion
                        if (t.getRawClass() == Long.class) {
                            return true;
                        }
                        return !t.isFinal(); // includes Object.class
                    default:
                        // case JAVA_LANG_OBJECT:
                        return (t.getRawClass() == Object.class);
                }
            }
        };
        mapTyper.init(JsonTypeInfo.Id.CLASS, null);
        mapTyper.inclusion(As.WRAPPER_ARRAY);
        mapObjectMapper.setDefaultTyping(mapTyper);

        // warm up codec
        try {
            byte[] s = mapObjectMapper.writeValueAsBytes(1);
            mapObjectMapper.readValue(s, Object.class);
        }
        catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }
```

That led to the following
<pre>
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Unexpected token (START_OBJECT), expected START_ARRAY: need JSON Array to contain As.WRAPPER_ARRAY type information for class java.lang.Object
at [Source: io.netty.buffer.ByteBufInputStream@2474df51; line: 1, column: 1]
at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:270)
at com.fasterxml.jackson.databind.DeserializationContext.wrongTokenException(DeserializationContext.java:1389)
at com.fasterxml.jackson.databind.DeserializationContext.reportWrongTokenException(DeserializationContext.java:1210)
at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._locateTypeId(AsArrayTypeDeserializer.java:137)
at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._deserialize(AsArrayTypeDeserializer.java:96)
at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer.deserializeTypedFromAny(AsArrayTypeDeserializer.java:71)
at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserializeWithType(UntypedObjectDeserializer.java:553)
at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:63)
at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3814)
at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2924)
at org.redisson.codec.JsonJacksonCodec$2.decode(JsonJacksonCodec.java:81)
at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:283)
at org.redisson.client.handler.CommandDecoder.decodeList(CommandDecoder.java:313)
at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:302)
at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:127)
at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:367)
at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
at java.lang.Thread.run(Thread.java:748)
</pre>

Not sure if this is the correct way to set the codec. I will have to take a deeper look tomorrow. 
```java
Config config = new Config();
config.setCodec(myRedissonCodec);
.
.
.
Redisson.create(config)
```

**Comment by earmenda** [comment]
Found a related question from 3 years ago. Looks like the same exact stack trace. 

https://stackoverflow.com/questions/29839468/jackson-serialization-deserialization-of-java-util-enummap-errors

**Comment by ieugen** [comment]
Hi @earmenda , 

Thanks for the solution. I'm also struggling with the fact that Redisson deceided to force the inclusion of data types inside the json file.

I get this in Redis instead of a simple key value object. 
It's aslo no simple way to eliminate the JsonType metadata. 

To me it seems like a very poor design decision on Redisson. At most, this feature should be optional. 
Plain JSON makes communicating between projects much easier. 
````
1) "{\"@class\":\"java.util.AbstractMap$SimpleEntry\",\"1\":\"STARTED\"}"
2) "{\"@class\":\"java.util.AbstractMap$SimpleEntry\",\"2\":\"DECLINED\"}"
3) "{\"@class\":\"java.util.AbstractMap$SimpleEntry\",\"3\":\"SUBMITTED\"}"
````

**Comment by mrniko** [comment]
@ieugen 
> It's aslo no simple way to eliminate the JsonType metadata.

you can extend `JsonJacksonCodec` and override `initTypeInclusion` method to avoid this

**Comment by ieugen** [comment]
Hi @mrniko , 

I'm not a Jacskon guru and I don't think I need to be one to operate Redisson to persist a class in a queue. Redisson looks very promising with the features it has. 

I did override `JsonJacksonCodec` and now I get this. I don't know how to move past it since there isn't a way to provide type information to the queue or a way to ask for a specific object type.

````
java.lang.ClassCastException: java.util.LinkedHashMap cannot be cast to platform.datacollection.events.AssessmentStatusChangedEvent
````

**Comment by DALDEI** [comment]
See related #1740

**Comment by mrniko** [comment]
@earmenda 

Unable to reproduce this with the latest version of Redisson. According to follow comment https://github.com/FasterXML/jackson-databind/issues/1932#issuecomment-370663371 the issue has been resolved in Jackson 2.9.5
